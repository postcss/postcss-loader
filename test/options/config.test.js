import path from 'path';

import {
  compile,
  getCompiler,
  getErrors,
  getCodeFromBundle,
  getWarnings,
} from '../helpers/index';

const testDirectory = path.resolve(__dirname, '../fixtures', 'config-autoload');

describe('"config" option', () => {
  it('should work Config - false', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: false,
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Config - true', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: true,
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Config - "string"', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: path.resolve(
          __dirname,
          '../fixtures/config-scope/css/custom.config.js'
        ),
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Config - "string" with relative path', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: 'test/fixtures/config-scope/css/custom.config.js',
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Config - "string" with path directory', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: 'test/fixtures/config-scope',
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Config - Object - path file', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: {
          path: path.resolve(
            __dirname,
            '../fixtures/config-scope/css/custom.config.js'
          ),
        },
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Config - {Object}', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {});
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Config - Path - {String}', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: { path: 'test/fixtures/config-scope/config/postcss.config.js' },
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Config - Context - {Object}', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: {
          path: 'test/fixtures/config-scope/config/postcss.config.js',
          ctx: { plugin: true },
        },
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work Config – Context – Loader {Object}', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        config: {
          path: 'test/fixtures/config-scope/config/context/postcss.config.js',
        },
      },
    });
    const stats = await compile(compiler);

    const { assets } = stats.compilation;

    const asset = 'asset.txt';

    expect(asset in assets).toBeTruthy();
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work postcss.config.js - {Object} - Process CSS', async () => {
    const compiler = getCompiler('./config-autoload/js/object/index.js', {
      postcssOptions: {
        config: {
          path: path.resolve(testDirectory, 'js/object'),
          ctx: { parser: false, syntax: false },
        },
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('index.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work postcss.config.js - {Array} - Process CSS', async () => {
    const compiler = getCompiler('./config-autoload/js/array/index.js', {
      postcssOptions: {
        config: {
          path: path.resolve(testDirectory, 'js/array'),
          ctx: { parser: false, syntax: false },
        },
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('index.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work package.json - {Object} - Process CSS', async () => {
    const compiler = getCompiler('./config-autoload/pkg/index.js', {
      postcssOptions: {
        config: {
          path: path.resolve(testDirectory, 'pkg'),
        },
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('index.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should throw an error when unresolved config ', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: path.resolve(
          __dirname,
          '../fixtures/config-scope/css/unresolve.js'
        ),
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats, true)).toMatchSnapshot('errors');
  });

  it('should throw an error on the invalid config ', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: path.resolve(
          __dirname,
          '../fixtures/config-scope/css/invalid.config.js'
        ),
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats, true)).toMatchSnapshot('errors');
  });

  it('should work if the "config" options is not specified', async () => {
    const compiler = getCompiler('./css/index.js', {});
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with the "postcssOptions" option', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: true,
        plugins: ['postcss-rtl'],
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });
});
