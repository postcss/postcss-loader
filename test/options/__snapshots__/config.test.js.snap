// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"config" option should throw an error on the invalid config : errors 1`] = `
Array [
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
Error: invalid postcss config",
]
`;

exports[`"config" option should throw an error on the invalid config : warnings 1`] = `Array []`;

exports[`"config" option should throw an error when unresolved config : errors 1`] = `
Array [
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
Error: No PostCSS Config found in: /test/fixtures/config-scope/css/unresolve.js",
]
`;

exports[`"config" option should throw an error when unresolved config : warnings 1`] = `Array []`;

exports[`"config" option should work Config - "string" with path directory: css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work Config - "string" with path directory: errors 1`] = `Array []`;

exports[`"config" option should work Config - "string" with path directory: warnings 1`] = `Array []`;

exports[`"config" option should work Config - "string" with relative path: css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work Config - "string" with relative path: errors 1`] = `Array []`;

exports[`"config" option should work Config - "string" with relative path: warnings 1`] = `Array []`;

exports[`"config" option should work Config - "string": css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work Config - "string": errors 1`] = `Array []`;

exports[`"config" option should work Config - "string": warnings 1`] = `Array []`;

exports[`"config" option should work Config - {Object}: css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work Config - {Object}: errors 1`] = `Array []`;

exports[`"config" option should work Config - {Object}: warnings 1`] = `Array []`;

exports[`"config" option should work Config - Context - {Object}: css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work Config - Context - {Object}: errors 1`] = `Array []`;

exports[`"config" option should work Config - Context - {Object}: warnings 1`] = `Array []`;

exports[`"config" option should work Config - Object - path file: css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work Config - Object - path file: errors 1`] = `Array []`;

exports[`"config" option should work Config - Object - path file: warnings 1`] = `Array []`;

exports[`"config" option should work Config - Path - {String}: css 1`] = `
"a { color: black }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work Config - Path - {String}: errors 1`] = `Array []`;

exports[`"config" option should work Config - Path - {String}: warnings 1`] = `Array []`;

exports[`"config" option should work Config - false: css 1`] = `
"a { color: black }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work Config - false: errors 1`] = `Array []`;

exports[`"config" option should work Config - false: warnings 1`] = `Array []`;

exports[`"config" option should work Config - true: css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work Config - true: errors 1`] = `Array []`;

exports[`"config" option should work Config - true: warnings 1`] = `Array []`;

exports[`"config" option should work Config – Context – Loader {Object}: errors 1`] = `Array []`;

exports[`"config" option should work Config – Context – Loader {Object}: warnings 1`] = `Array []`;

exports[`"config" option should work if the "config" options is not specified: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"config" option should work if the "config" options is not specified: errors 1`] = `Array []`;

exports[`"config" option should work if the "config" options is not specified: warnings 1`] = `Array []`;

exports[`"config" option should work package.json - {Object} - Process CSS: css 1`] = `
".import {
  color: red;
}

.test {
  color: blue;
}
"
`;

exports[`"config" option should work package.json - {Object} - Process CSS: errors 1`] = `Array []`;

exports[`"config" option should work package.json - {Object} - Process CSS: warnings 1`] = `Array []`;

exports[`"config" option should work postcss.config.js - {Array} - Process CSS: css 1`] = `
".import {
  color: red;
}

.test {
  color: cyan;
}
"
`;

exports[`"config" option should work postcss.config.js - {Array} - Process CSS: errors 1`] = `Array []`;

exports[`"config" option should work postcss.config.js - {Array} - Process CSS: warnings 1`] = `Array []`;

exports[`"config" option should work postcss.config.js - {Object} - Process CSS: css 1`] = `
".import {
  color: red;
}

.test {
  color: blue;
}
"
`;

exports[`"config" option should work postcss.config.js - {Object} - Process CSS: errors 1`] = `Array []`;

exports[`"config" option should work postcss.config.js - {Object} - Process CSS: warnings 1`] = `Array []`;

exports[`"config" option should work with the "postcssOptions" option: css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

[dir=ltr] .foo { float: right;
}

[dir=rtl] .foo {
  float: left;
}
"
`;

exports[`"config" option should work with the "postcssOptions" option: errors 1`] = `Array []`;

exports[`"config" option should work with the "postcssOptions" option: warnings 1`] = `Array []`;
